File: prozesse_und_threads/vorgabe/07_process_threads_difference.py
import threading
import multiprocessing

N = 2
# TODO

def dummy_task(name):
    print(f"Start {name}")
    # TODO
    print(f"End {name}")

if __name__ == '__main__':
    threads = [threading.Thread(target=dummy_task, args=(f"Thread-{i+1}",)) for i in range(N)]
    for t in threads: t.start()
    for t in threads: t.join()

    processes = [multiprocessing.Process(target=dummy_task, args=(f"Process-{i+1}",)) for i in range(N)]
    for p in processes: p.start()
    for p in processes: p.join()

File: prozesse_und_threads/vorgabe/07_process_threads_difference.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <unistd.h>
#include <sys/wait.h>

int counter = 0;

void dummy_task(const std::string &name)
{
    std::cout << "Start " << name << std::endl;
    for (int i = 0; i < 10; i++)
    {
        counter += 1;
        std::cout << name << ": " << counter << std::endl;
    }
    std::cout << "End " << name << std::endl;
}

int main()
{
    std::thread thread1(dummy_task, "Thread-1");
    std::thread thread2(dummy_task, "Thread-2");
    thread1.join();
    thread2.join();

    counter = 0;

    pid_t pid = fork();
    if (pid == 0)
    {
        // Kindprozess
        dummy_task("Process-2");
    }
    else if (pid > 1)
    {
        dummy_task("Process-1");
        int status;
        wait(&status);
    }

    return 0;
}
